openedx/core/djangoapps/credit/services.py

```diff
@@ -100,11 +100,18 @@ def get_credit_state(self, user_id, course_key_or_id, return_course_info=False):
         }
 
         if return_course_info:
-            course_overview = CourseOverview.get_from_id(course_key)
-            result.update({
-                'course_name': course_overview.display_name,
-                'course_end_date': course_overview.end,
-            })
+            try:
+                course_overview = CourseOverview.get_from_id(course_key)
+                result.update({
+                    'course_name': course_overview.display_name,
+                    'course_end_date': course_overview.end,
+                })
+            except CourseOverview.DoesNotExist:
+                log.exception(
+                    "Could not get name and end_date for course %s, This happened because we were unable to "
+                    "get/create CourseOverview object for the course. It's possible that the Course has been deleted.",
+                    course_key_or_id,
+                )
```

Comment:
What will happen if this method returns with a `course_name` and `course_end_date`? Will we be causing errors up the stack?

Reply:
I believe you meant **without** `a course_name and course_end_date` above.

I believe in this case we are majorly relying on the existing coverage. However, On a long shot, I looked around the usages of `get_credit_state` to check an impact that might have. Mentioning them below:

**In Platform:**
1) There is only [this](https://github.com/openedx/edx-platform/blob/master/common/lib/xmodule/xmodule/seq_module.py#L945) usage of it in the platform and it doesn't pass `return_course_info=True` it might never end up in this block.

**In edx-proctoring** (The usage where `return_course_info=True`, we don't need to check if it's False)
1) The impact [first usage](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L1638-L1661) will have is that in the credits update email the course name will be sent as `your course` instead of ac actual course name. In addition, the attempt status is set to `Submitted` so this exam/attempt will be treated as done in the future.
2) I have some concerns about the [second usage](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2380-L2393) though. This is the usage that purely deals with the `course_end_date` and sets an exam status eligibility or expiry if an [attempt isn't found](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2388-L2389) for an exam. the relevant function([get_attempt_status_summary](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2327)) which is further used in some placed in edx-platform in milestones and special exams. Since we are not going to have a course and based on that I think we can send any time in the past for `course_end_date` so that if there is no course this would be treated as expired. (Or do we have any other options if we don't have a course at all).

Reply:
@pdpinch I believe you meant `without a course_name and course_end_date? `?

I think in that case our best shot is the existing coverage. However, in a long shot, I looked for the usages of `get_credit_state` and there are a few for which I'll present my opinion:
1) There is only [this]() usage in the `edx-platform` itself but since it doesn't pass a `return_course_info/True` flag it will never come in the block we changed.

Reply:
Is it possible to add a test to cover the second usage when the course no longer exists?

Reply:
>Is it possible to add a test to cover the second usage when the course no longer exists?

@pdpinch I'm sorry I didn't quite understand, can you probably explain a little more about which part we will be covering?
Although, there are a few things I can add for reference.

Currently, What's going to happen in the case when there is no course is that either the status is going to be set to [current exam attempt status](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2389) or [eligible](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2393) because `due_date_is_passed` is always going to be `False` when the course no longer exists.
And the function [get_attempt_summary_status](https://github.com/openedx/edx-proctoring/blob/master/edx_proctoring/api.py#L2327) is called in [special_exams](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content/learning_sequences/api/processors/special_exams.py#L58-L64), [milestones](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/milestones.py#L118-L124) and [module_renderer](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L254-L268) all of which are using try/except and the comments above them give an impression that it might never be a case for us?

Reply:
Thank you again for the context, @arslanashraf7!
So sorry for making you repeat yourself in the other comment, this problem is kind of far-fetched! I'm glad you found a workaround.

I have a few questions though:

* Did you resolve the concerns about the second usage in edx-proctoring?
* About that. If there's no course, are the effects of the ghost attempt relevant?
* Do you have any other side effects in mind do you want to discuss?

I tested this a couple of times and works like a charm.

Reply:
Great! Thanks. What concerns me the most about 1) is how misleading is the eligible status in this situation. It seems like the safest answer is to just set the `course_end_date` to a date in the past. What do you think?

Reply:
I agree, The safest would be to go with some past `course_end_date` and allow the attempt status to expire.
I think you're good with it too but just to reconfirm, I'll make this change in the PR for the second case with an additional test. What past time do you suggest (A day in the past, An hour, etc..)?

Reply:
Well as [this](https://github.com/openedx/edx-proctoring/blob/58b5fa227df004009fec3b1bec800b5f0d22e0f8/edx_proctoring/utils.py#L242) function just checks a datetime less or equal than the current time, it would be enough just to use a `timedelta` of an hour :thinking:,  but either option you choose, let's leave a comment explaining why are we setting the course_end_date to that value :smile:

Reply:
@mariajgrimaldi The above has been incorporated and I've also updated the test.